### PR TITLE
Support self-relay with IGP enforced on warp routes

### LINKED ISSUE #3525: Support self-relay with IGP enforced on warp routes
## Problem

Warp routes have a [hook](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/df34198d4c230f462673d009646986cf275869fa/solidity/contracts/client/MailboxClient.sol#L21) configurable by the `owner` but do not allow users to override this hook or metadata.

## Solution

Overload [`transferRemote`](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/b14f997810ebd7dbdff2ac6622a149ae77010ae3/solidity/contracts/token/libs/TokenRouter.sol#L53) with a `hook` and `metadata`.

## Nice to Have

- [ ] Support a `relayer` address override option in the IGP metadata and on the contract

### PR DESCRIPTION
**Pull request by ltyu**
### Description
Adds transferRemote overload 

### Related issues

- Fixes #3525 

### Backward compatibility
Yes

### Testing
Unit Tests

### CODE REVIEW

File: solidity/contracts/token/libs/TokenRouter.sol [hunk 1]
```diff
@@ -84,6 +85,69 @@ abstract contract TokenRouter is GasRouter {
...
+    ) internal returns (bytes32 messageId) {
+        bytes memory metadata = _transferFromSender(_amountOrId);
+        messageId = _dispatch(
+            _destination,
+            _recipient,
+            _gasPayment,
+            TokenMessage.format(_recipient, _amountOrId, metadata),
+            _hookMetadata,
+            _hook
+        );
```
Comment (yorhodes) [inline solidity/contracts/token/libs/TokenRouter.sol:147]: would prefer to dedupe this with the existing _transferRemote if possible

File: solidity/test/token/HypERC20.t.sol [hunk 2]
```diff
@@ -185,6 +192,47 @@ abstract contract HypTokenTest is Test {
...
+        /// @dev Using this test wold be ideal, but vm.expectCall with nested functions more than 1 level deep is broken
+        /// In other words, the call graph of Route.transferRemote() -> Mailbox.dispatch() -> Hook.postDispatch() does not work with expectCall
+        // vm.expectCall(
+        //     address(noopHook),
+        //     abi.encodeCall(
+        //         IPostDispatchHook.postDispatch,
+        //         (bytes(""), outboundMessage)
+        //     )
+        // );
+        /// @dev Also, using expectedCall with Mailbox.dispatch() won't work either because overloaded function selection is broken, see https://github.com/ethereum/solidity/issues/13815
```
Comment (yorhodes) [inline solidity/test/token/HypERC20.t.sol:233]: brutal

**Review by yorhodes — COMMENTED** [review]
nice quick one

**Review by yorhodes — APPROVED** [review]

File: solidity/test/token/HypERC20.t.sol (hunk 2 above)
Comment (ltyu) [reply solidity/test/token/HypERC20.t.sol:233]: 😭

File: solidity/contracts/token/libs/TokenRouter.sol (hunk 1 above)
Comment (ltyu) [reply solidity/contracts/token/libs/TokenRouter.sol]: Fixed

**Pull request merged**